src/Cardano/Address/Style/Shelley.hs

```diff
@@ -169,21 +218,68 @@ instance SoftDerivation Shelley where
+          -- specification - Section 3.2.3. What is important here is that the
+          -- address is composed of discrimination byte and 28 bytes hashed public key.
+          -- Moreover, it was decided that first 4 bits for enterprise address
+          -- will be `0110`. The next for bits for network discriminator are to
+          -- be established soon. TO_DO : change `firstbyte` as network
+          -- discriminant in CDDL spec is known.
+          firstByte = case discrimination of
+              RequiresNoMagic  -> 0x60
+              RequiresMagic (ProtocolMagic _pm) -> 0x61
+          expectedLength = 1 + publicKeyHashSize
+
+instance DelegationAddress Shelley where
+    delegationAddress discrimination paymentKey stakingKey = unsafeMkAddress $
+        invariantSize expectedLength $ BL.toStrict $ runPut $ do
+            putWord8 firstByte
+            mapM_ (putByteString . keyToHash) [paymentKey, stakingKey]
+      where
+          -- we use here the fact that payment address stands for what is named
+          -- as base address - refer to delegation specification - Section 3.2.1.
+          -- What is important here is that the address is composed of discrimination
+          -- byte and two 28 bytes hashed public keys, one for payment key and the
+          -- other for staking/reword key.
+          -- Moreover, it was decided that first 4 bits for enterprise address
+          -- will be `0000`. The next for bits for network discriminator are to
+          -- be established soon. TO_DO : change `firstbyte` as network
+          -- discriminant in CDDL spec is known.
+          firstByte = case discrimination of
+              RequiresNoMagic  -> 0x00
+              RequiresMagic (ProtocolMagic _pm) -> 0x01
+          expectedLength = 1 + 2*publicKeyHashSize
+
+invariantSize :: HasCallStack => Int -> ByteString -> ByteString
+invariantSize expectedLength bytes
+    | BS.length bytes == expectedLength = bytes
+    | otherwise = error
+      $ "length was "
+      ++ show (BS.length bytes)
+      ++ ", but expected to be "
+      ++ (show expectedLength)
+
+keyToHash :: Shelley depth XPub -> ByteString
+keyToHash = BA.convert .
+    hash @_ @Blake2b_224 .
+    xpubToBytesWithoutChainCode .
+    getKey
+
 {-------------------------------------------------------------------------------
                                  Key generation
 -------------------------------------------------------------------------------}
 
--- | Size, in bytes, of a public key (without chain code)
-publicKeySize :: Int
-publicKeySize = 32
-
--- Serialized length in bytes of a Single Address
-addrSingleSize :: Int
-addrSingleSize = 1 + publicKeySize
-
--- Serialized length in bytes of a Grouped Address
-addrGroupedSize :: Int
-addrGroupedSize = addrSingleSize + publicKeySize
+-- | Size, in bytes, of a hash of public key (without chain code)
+publicKeyHashSize :: Int
+publicKeyHashSize = 28
```

Comment:
Note that instead of hard-coding the constant here, we could simply use: https://hackage.haskell.org/package/cryptonite-0.26/docs/Crypto-Hash-IO.html#v:hashDigestSize

```hs
hashDigestSize Blake2b_224
28 
```

Reply:
done in https://github.com/input-output-hk/cardano-addresses/pull/20/commits/b97b25b7bbf9c5f2130b016a8e88457325063549